File: harvard_cs50/week_1/p_set1/credit_bad/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long card_number = get_long("Please enter your credit card number: ");
    int card_total = 0;
    for (long i = 10; i < 1000000000000000000; i = i * 100)
    {
        long card = card_number / i % 10;
        int num_first = card * 2;
        int num_last = (card * 2) % 10;
        if (num_first > 10)
        {
            while(num_first >= 10)
                {
                num_first = num_first / 10;
                }
        }
        else
        {
            num_first = 0;
        }
        card_total = card_total + num_first + num_last;
    }
    for (long i = 1; i < 10000000000000000; i = i * 100)
    {
        long card = card_number / i % 10;
        card_total = card_total + card;
    }
    if (card_total % 10 == 0)
    {
        while (card_number >= 10)
            {
                card_number = card_number / 10;
            }
        printf("%li", card_number);
        if (card_number == 3)
            {
                printf("AMEX\n");
            }
        else if (card_number == 5)
            {
                printf("MASTERCARD\n");
            }
        else if (card_number == 4)
            {
                printf("VISA\n");
            }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: harvard_cs50/week_6/pset_6/sentimental-credit/credit.py
from cs50 import get_int

# checking validity of card


def is_valid(number):
    # list comprehension changing number into a list of strings
    number = [int(x) for x in str(number)]
    # number_1 is now a list of every second element times to starting from the end
    number_1 = [str(x * 2) for x in number[-2::-2] if x != 0]
    number_2 = [x for x in number[-1::-2]]

    for i in range(len(number_1)):
        if len(number_1[i]) > 1:
            number_1[i] = int(number_1[i][0]) + int(number_1[i][1])
    number_1 = [int(x) for x in number_1]

    if (sum(number_1) + sum(number_2)) % 10 == 0:
        return True
    return False

# checking type of card


def check_type(number):
    number = str(number)
    if number[0:2] in ['34', '37'] and len(number) == 15:
        print("AMEX")
    elif number[0:2] in ['51', '52', '53', '54', '55'] and len(number) == 16:
        print("MASTERCARD")
    elif number[0] == '4' and len(number) == 16 or len(number) == 13:
        print("VISA")
    else:
        print("INVALID")

# Main function


def main():
    card_number = get_int("Please Input Card Number: ")

    if is_valid(card_number):
        print(check_type(card_number))
    else:
        print("INVALID")


main()